sw/device/silicon_creator/lib/drivers/retention_sram.c

```diff
@@ -0,0 +1,30 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sram_ctrl_regs.h"  // Generated.
+
+rom_error_t retention_sram_scramble(void) {
+  const uint32_t kBase = TOP_EARLGREY_SRAM_CTRL_RET_AON_BASE_ADDR;
+
+  // Check that control register writes are enabled.
+  if (!bitfield_bit32_read(
+          abs_mmio_read32(kBase + SRAM_CTRL_CTRL_REGWEN_REG_OFFSET),
+          SRAM_CTRL_CTRL_REGWEN_CTRL_REGWEN_BIT)) {
+    return kErrorRetSramLocked;
+  }
+
+  // Request the renewal of the scrambling key and initialization to random
+  // values.
+  uint32_t ctrl = 0;
+  ctrl = bitfield_bit32_write(ctrl, SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true);
+  ctrl = bitfield_bit32_write(ctrl, SRAM_CTRL_CTRL_INIT_BIT, true);
+  abs_mmio_write32(kBase + SRAM_CTRL_CTRL_REG_OFFSET, ctrl);
```

Comment:
I checked the design, and it looks like setting both bits simultaneously should work fine.
TL-UL accesses to SRAM are stalled until the key has been obtained and the SRAM has been initialized.
So it is technically OK to NOT spinwait on the `SCR_KEY_VALID` bit.

As for the `SCR_KEY_SEED_VALID` bit: this will only be set to 1 if the key seeds have actually been provisioned.
So this field will behave differently depending on the provisioning state.
The obtained scrambling keys will however always be ephemeral, no matter whether the key seeds have been provisioned or not (the seeds are just a parameterization component of the key derivation function in addition to entropy fetched from EDN). It is OK to not explicitly check this.

See also the discussion here: https://github.com/lowRISC/opentitan/issues/7059#issuecomment-867078425